src/aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -192,74 +198,134 @@ def setup(self):
+        # If no work chain has been run, we should at least run one
+        if 'base_relax_workchains' not in self.ctx:
+            return True
+
+        # If meta convergence is switched off we are done
+        if not self.ctx.meta_convergence:
+            self.report('Meta-convergence disabled: workchain completed after a single iteration.')
+            return False
+
+        # Stop if the maximum number of meta iterations has been reached
+        if self.ctx.iteration == self.inputs.max_meta_convergence_iterations.value:
+            self.report('Maximum number of meta convergence iterations reached.')
+            return False
+
+        base_relax_workchain = self.ctx.base_relax_workchains[-1]
+
+        # If the last work chain still found Pulay stresses in the final SCF, continue
+        pulay_exit_status = PwCalculation.exit_codes.ERROR_IONIC_CONVERGENCE_REACHED_EXCEPT_IN_FINAL_SCF.status
+        if base_relax_workchain.exit_status == pulay_exit_status:
+            self.report('Pulay stresses still present, running another geometry optimizatio.')
+            return True
+
+        # If the kpoints are defined as a density, make sure the kpoints mesh is the same for the new structur
+        if 'kpoints_distance' in self.ctx.relax_inputs:
+            input_kpts_mesh, _ = base_relax_workchain.base.links.get_outgoing(
+                link_label_filter='iteration_01'
+            ).first().node.inputs.kpoints.get_kpoints_mesh()
+            inputs_create_kpoints = {
+                'structure': base_relax_workchain.outputs.output_structure,
+                'distance': self.ctx.relax_inputs.kpoints_distance,
+                'force_parity': self.ctx.relax_inputs.get('kpoints_force_parity', orm.Bool(False)),
+                'metadata': {
+                    'store_provenance': False
+                }
+            }
+            new_kpts_mesh, _ = create_kpoints_from_distance(**inputs_create_kpoints).get_kpoints_mesh()
+            if tuple(input_kpts_mesh) != tuple(new_kpts_mesh):
+                self.report(
+                    'k-points mesh changed for specified `kpoints_distance` due to a change in the unit cell. '
+                    'Running another geometry optimization with new mesh.'
+                )
+                return True
+
+        self.report(f'Work chain completed after {self.ctx.iteration} iterations.')
+        return False
 
     def run_relax(self):
         """Run the `PwBaseWorkChain` to run a relax `PwCalculation`."""
         self.ctx.iteration += 1
-
         inputs = self.ctx.relax_inputs
-        inputs.pw.structure = self.ctx.current_structure
 
-        # If one of the nested `PwBaseWorkChains` changed the number of bands, apply it here
+        if 'base_init_relax_workchain' in self.ctx:
+            inputs.pw.structure = self.ctx.base_init_relax_workchain.outputs.output_structure
+        else:
+            inputs.pw.structure = self.ctx.current_structure
```

Comment:
This is not necessary because `self.ctx.current_structure` is correctly set in `inspect_init_relax`